File: sdk/runanywhere-swift/Sources/RunAnywhere/Features/VAD/Services/VADComponent.swift
//
//  VADComponent.swift
//  RunAnywhere SDK
//
//  Component wrapper for VAD to integrate with the component lifecycle system
//

@preconcurrency import AVFoundation
import Foundation

/// Voice Activity Detection component following the clean architecture
/// Extends BaseComponent to integrate with the SDK's component lifecycle system
@MainActor
public final class VADComponent: BaseComponent<VADServiceWrapper>, @unchecked Sendable {

    // MARK: - Properties

    public override static var componentType: SDKComponent { .vad }

    private let vadConfiguration: VADConfiguration
    private var isPaused: Bool = false
    private let logger = SDKLogger(category: "VADComponent")
    private let analyticsService = VADAnalyticsService()

    // MARK: - Initialization

    public init(configuration: VADConfiguration, serviceContainer: ServiceContainer? = nil) {
        self.vadConfiguration = configuration
        super.init(configuration: configuration, serviceContainer: serviceContainer)
    }

    // MARK: - Service Creation

    public override func createService() async throws -> VADServiceWrapper {
        // Create default SimpleEnergyVADService
        // Note: VAD currently uses SimpleEnergyVADService directly
        // Future enhancement: Support VAD provider pattern via ModuleRegistry like STT/TTS
        logger.info("Creating SimpleEnergyVADService")
        let vadService = SimpleEnergyVADService(
            sampleRate: vadConfiguration.sampleRate,
            frameLength: vadConfiguration.frameLength,
            energyThreshold: vadConfiguration.energyThreshold
        )
        try await vadService.initialize()

        // Wrap the service
        let wrapper = VADServiceWrapper(vadService)
        return wrapper
    }

    public override func performCleanup() async throws {
        // No specific cleanup needed for VAD
    }

    // MARK: - Helper Methods

    private var vadService: (any VADService)? {
        return service?.wrappedService
    }

    // MARK: - Pause and Resume

    /// Pause VAD processing
    public func pause() {
        isPaused = true
        vadService?.pause()
    }

    /// Resume VAD processing
    public func resume() {
        isPaused = false
        vadService?.resume()
    }

    // MARK: - Public API

    /// Detect speech in audio buffer
    public func detectSpeech(in buffer: AVAudioPCMBuffer) async throws -> VADOutput {
        try ensureReady()

        guard let vadService = vadService else {
            throw VADError.serviceNotAvailable
        }

        // Apply threshold if configured
        if vadConfiguration.energyThreshold != vadService.energyThreshold {
            vadService.energyThreshold = vadConfiguration.energyThreshold
        }

        // Process buffer
        vadService.processAudioBuffer(buffer)

        // Get current state
        let isSpeechDetected = vadService.isSpeechActive

        // Track detection
        await analyticsService.trackDetection(
            isSpeechDetected: isSpeechDetected,
            energyLevel: vadService.energyThreshold
        )

        return VADOutput(
            isSpeechDetected: isSpeechDetected,
            energyLevel: vadService.energyThreshold
        )
    }

    /// Detect speech in audio samples
    public func detectSpeech(in samples: [Float]) async throws -> VADOutput {
        try ensureReady()

        guard let vadService = vadService else {
            throw VADError.serviceNotAvailable
        }

        // Process samples and get result
        let isSpeechDetected = vadService.processAudioData(samples)

        // Track detection
        await analyticsService.trackDetection(
            isSpeechDetected: isSpeechDetected,
            energyLevel: vadService.energyThreshold
        )

        return VADOutput(
            isSpeechDetected: isSpeechDetected,
            energyLevel: vadService.energyThreshold
        )
    }

    /// Process VAD input (supports both buffer and samples)
    public func process(_ input: VADInput) async throws -> VADOutput {
        // Apply threshold override if provided
        if let threshold = input.energyThresholdOverride,
           let vadService = vadService {
            vadService.energyThreshold = threshold
        }

        // Process based on input type
        if let buffer = input.buffer {
            return try await detectSpeech(in: buffer)
        } else if let samples = input.audioSamples {
            return try await detectSpeech(in: samples)
        } else {
            throw VADError.invalidInput(reason: "VADInput must contain either buffer or audioSamples")
        }
    }

    /// Process audio stream
    public func processAudioStream<S: AsyncSequence>(_ stream: S) -> AsyncThrowingStream<VADOutput, Error>
    where S.Element == AVAudioPCMBuffer {
        AsyncThrowingStream { continuation in
            Task {
                do {
                    for try await buffer in stream {
                        let output = try await detectSpeech(in: buffer)
                        continuation.yield(output)
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    /// Reset VAD state
    public func reset() {
        vadService?.reset()
    }

    /// Set speech activity callback
    public func setSpeechActivityCallback(_ callback: @escaping (SpeechActivityEvent) -> Void) {
        vadService?.onSpeechActivity = { [weak self] event in
            callback(event)

            // Track speech activity events
            Task { [weak self] in
                guard let self = self else { return }
                switch event {
                case .started:
                    await self.analyticsService.trackSpeechActivityStarted()
                case .ended:
                    // We don't have duration info from the event, pass 0
                    await self.analyticsService.trackSpeechActivityEnded(duration: 0)
                }
            }
        }
    }

    /// Start VAD processing
    public func start() {
        vadService?.start()
    }

    /// Stop VAD processing
    public func stop() {
        vadService?.stop()
    }

    /// Get the underlying VAD service
    public func getService() -> VADService? {
        return vadService
    }

    /// Start calibration of the VAD
    public func startCalibration() async throws {
        try ensureReady()

        guard let vadService = vadService else {
            throw VADError.serviceNotAvailable
        }

        // These methods are specific to SimpleEnergyVADService
        guard let energyVAD = vadService as? SimpleEnergyVADService else {
            logger.warning("Calibration is only supported for SimpleEnergyVADService")
            return
        }

        let startTime = Date()
        let thresholdBefore = vadService.energyThreshold

        // Track calibration started
        await analyticsService.trackCalibrationStarted(currentThreshold: thresholdBefore)

        await energyVAD.startCalibration()

        let duration = Date().timeIntervalSince(startTime)
        let thresholdAfter = vadService.energyThreshold

        // Track calibration completed
        await analyticsService.trackCalibrationCompleted(
            thresholdBefore: thresholdBefore,
            thresholdAfter: thresholdAfter,
            samplesCollected: 20, // Based on SimpleEnergyVADService default
            duration: duration * 1000 // Convert to ms
        )
    }

    /// Get current VAD statistics for debugging
    public func getStatistics() -> VADStatistics? {
        guard let energyVAD = vadService as? SimpleEnergyVADService else {
            return nil
        }
        return energyVAD.getStatistics()
    }

    /// Set calibration parameters
    public func setCalibrationParameters(multiplier: Float) {
        guard let energyVAD = vadService as? SimpleEnergyVADService else {
            logger.warning("setCalibrationParameters is only supported for SimpleEnergyVADService")
            return
        }
        energyVAD.setCalibrationParameters(multiplier: multiplier)
    }

    /// Get analytics service for advanced tracking
    public func getAnalyticsService() -> VADAnalyticsService {
        return analyticsService
    }
}

File: sdk/runanywhere-kotlin/src/commonMain/kotlin/com/runanywhere/sdk/features/vad/VADComponent.kt
package com.runanywhere.sdk.features.vad

import com.runanywhere.sdk.core.ModuleRegistry
import com.runanywhere.sdk.core.capabilities.*
import com.runanywhere.sdk.data.models.SDKError
import com.runanywhere.sdk.models.enums.InferenceFramework
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

// MARK: - VAD Component

/**
 * Voice Activity Detection component following the clean architecture.
 * Integrates with VADAnalyticsService for event tracking matching iOS pattern.
 */
class VADComponent(
    configuration: VADConfiguration,
    private val analyticsService: VADAnalyticsService = VADAnalyticsService(),
) : BaseComponent<VADServiceWrapper>(configuration) {
    // MARK: - Properties

    override val componentType: SDKComponent = SDKComponent.VAD

    private val vadConfiguration: VADConfiguration = configuration

    /** Coroutine scope for analytics operations */
    private val analyticsScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

    // MARK: - Service Creation

    override suspend fun createService(): VADServiceWrapper {
        // Try to get a registered VAD provider from central registry
        val provider =
            ModuleRegistry.vadProvider(vadConfiguration.modelId)
                ?: run {
                    analyticsService.trackInitializationFailed(
                        error = "No VAD service provider registered",
                        framework = InferenceFramework.BUILT_IN,
                    )
                    throw SDKError.ComponentNotInitialized(
                        "No VAD service provider registered. Please register WebRTCVADServiceProvider.register()",
                    )
                }

        try {
            // Create service through provider
            val vadService = provider.createVADService(vadConfiguration)

            // Track successful initialization (using BUILT_IN for energy-based VAD)
            analyticsService.trackInitialized(framework = InferenceFramework.BUILT_IN)

            // Wrap the service
            return VADServiceWrapper(vadService)
        } catch (e: Exception) {
            analyticsService.trackInitializationFailed(
                error = e.message ?: "Unknown error",
                framework = InferenceFramework.BUILT_IN,
            )
            throw e
        }
    }

    override suspend fun performCleanup() {
        analyticsService.trackCleanedUp()
        service?.wrappedService?.cleanup()
    }

    // MARK: - Helper Methods

    private val vadService: VADService?
        get() = service?.wrappedService

    // MARK: - Public API

    /**
     * Process audio samples for voice activity detection - matching iOS detectSpeech(in: [Float])
     */
    fun processAudioChunk(audioSamples: FloatArray): VADOutput {
        ensureReady()

        val input =
            VADInput(
                audioSamples = audioSamples,
            )
        return process(input)
    }

    /**
     * Detect speech in audio samples - iOS-style method name
     */
    fun detectSpeech(audioSamples: FloatArray): VADOutput = processAudioChunk(audioSamples)

    /**
     * Process VAD input - supporting threshold override like iOS
     */
    fun process(input: VADInput): VADOutput {
        ensureReady()

        val service = vadService ?: throw SDKError.ComponentNotReady("VAD service not available")

        // Validate input
        input.validate()

        // Apply threshold override if provided (matching iOS pattern)
        val originalThreshold = service.energyThreshold
        input.energyThresholdOverride?.let { override ->
            service.energyThreshold = override
        }

        try {
            // Process audio chunk
            val result = service.processAudioChunk(input.audioSamples)

            // Calculate energy level (simple RMS)
            val energyLevel = calculateEnergyLevel(input.audioSamples)

            return VADOutput(
                isSpeechDetected = result.isSpeechDetected,
                energyLevel = energyLevel,
                confidence = result.confidence,
            )
        } finally {
            // Restore original threshold if it was overridden
            if (input.energyThresholdOverride != null) {
                service.energyThreshold = originalThreshold
            }
        }
    }

    /**
     * Stream VAD processing
     */
    fun streamProcess(audioStream: Flow<FloatArray>): Flow<VADOutput> =
        flow {
            ensureReady()

            audioStream.collect { audioSamples ->
                val output = processAudioChunk(audioSamples)
                emit(output)
            }
        }.catch { error ->
            throw VADError.ProcessingFailed(error)
        }

    /**
     * Process with speech segments detection.
     * Tracks speech start/end events for analytics matching iOS pattern.
     */
    fun detectSpeechSegments(
        audioStream: Flow<FloatArray>,
        onSpeechStart: () -> Unit = {},
        onSpeechEnd: () -> Unit = {},
    ): Flow<VADOutput> =
        flow {
            ensureReady()

            var isInSpeech = false
            var silenceFrames = 0
            // Use iOS-style hysteresis - 10 frames of silence to end speech
            val silenceFramesThreshold = 10

            audioStream.collect { audioSamples ->
                val output = processAudioChunk(audioSamples)

                when {
                    output.isSpeechDetected && !isInSpeech -> {
                        isInSpeech = true
                        silenceFrames = 0
                        // Track speech start for analytics
                        analyticsScope.launch { analyticsService.trackSpeechStart() }
                        onSpeechStart()
                    }

                    !output.isSpeechDetected && isInSpeech -> {
                        silenceFrames++
                        if (silenceFrames >= silenceFramesThreshold) {
                            isInSpeech = false
                            // Track speech end for analytics
                            analyticsScope.launch { analyticsService.trackSpeechEnd() }
                            onSpeechEnd()
                        }
                    }

                    output.isSpeechDetected && isInSpeech -> {
                        silenceFrames = 0
                    }
                }

                emit(output)
            }
        }

    /**
     * Reset VAD state
     */
    fun reset() {
        vadService?.reset()
    }

    /**
     * Set speech activity callback (matching iOS pattern)
     */
    fun setSpeechActivityCallback(callback: (SpeechActivityEvent) -> Unit) {
        vadService?.onSpeechActivity = callback
    }

    /**
     * Set audio buffer callback (matching iOS pattern)
     */
    fun setAudioBufferCallback(callback: (ByteArray) -> Unit) {
        vadService?.onAudioBuffer = callback
    }

    /**
     * Get service for compatibility
     */
    fun getService(): VADService? = vadService

    /**
     * Check if VAD is enabled
     */
    fun isEnabled(): Boolean = state == ComponentState.READY

    /**
     * Start VAD processing - matching iOS method.
     * Tracks started event for analytics.
     */
    fun start() {
        analyticsScope.launch { analyticsService.trackStarted() }
        vadService?.start()
    }

    /**
     * Stop VAD processing - matching iOS method.
     * Tracks stopped event for analytics.
     */
    fun stop() {
        analyticsScope.launch { analyticsService.trackStopped() }
        vadService?.stop()
    }

    /**
     * Pause VAD processing - matching iOS method.
     * Tracks paused event for analytics.
     */
    fun pause() {
        analyticsScope.launch { analyticsService.trackPaused() }
        // VAD service may not have pause, but we track the event
    }

    /**
     * Resume VAD processing - matching iOS method.
     * Tracks resumed event for analytics.
     */
    fun resume() {
        analyticsScope.launch { analyticsService.trackResumed() }
        // VAD service may not have resume, but we track the event
    }

    /**
     * Enable VAD (initialize if needed)
     */
    suspend fun enable() {
        if (state != ComponentState.READY) {
            initialize()
        }
    }

    /**
     * Disable VAD
     */
    suspend fun disable() {
        if (state == ComponentState.READY) {
            cleanup()
        }
    }

    /**
     * Process audio and keep pipeline warm
     */
    fun processAudio(audioData: ByteArray): ByteArray {
        // Simply pass through for now
        // In production, this would apply VAD filtering
        return audioData
    }

    // MARK: - Analytics

    /**
     * Get current VAD analytics metrics.
     * Matches iOS getAnalyticsMetrics() pattern.
     */
    suspend fun getAnalyticsMetrics(): VADMetrics = analyticsService.getMetrics()

    // MARK: - Private Helpers

    private fun calculateEnergyLevel(audioSamples: FloatArray): Float {
        if (audioSamples.isEmpty()) return 0f

        var sum = 0.0
        for (sample in audioSamples) {
            sum += sample * sample
        }

        return kotlin.math.sqrt(sum / audioSamples.size).toFloat()
    }
}
